docs/SPEC.md

```diff
@@ -1284,6 +1289,73 @@ Examples using known start and stop dates:
 
+#### Location support
+
+Intervals have location support for changing offsets by default.
+When a location has been specified either with the `location` option or the `location` parameter, a change in offset is used to either expand or shrink an interval.
+The most common example is daylight savings time switches.
+When daylight savings time starts, the zone offset shifts west, becoming a more negative number.
+When this happens, the hour that was lost does not exist and the current window is shortened to accomodate this change.
+When daylight savings time ends, the zone offset shifts east, becoming a more positive number.
+When this happens, one of the hours happens twice and so the interval becomes longer.
+
+Examples:
+
+    intervals(every:1d, location: "America/Denver")
+    // [2019-03-09T00:00:00-07:00, 2019-03-10T00:00:00-07:00)
+    // [2019-03-10T00:00:00-07:00, 2019-03-11T00:00:00-06:00)
+    // [2019-03-11T00:00:00-06:00, 2019-03-12T00:00:00-06:00)
+    // ... skipped ...
+    // [2019-11-02T00:00:00-06:00, 2019-11-03T00:00:00-06:00)
+    // [2019-11-03T00:00:00-06:00, 2019-11-04T00:00:00-07:00)
+    // [2019-11-04T00:00:00-07:00, 2019-11-05T00:00:00-07:00)
+
+    intervals(every:2h, location: "America/Denver")
+    // [2019-03-10T00:00:00-07:00, 2019-03-10T03:00:00-06:00)
+    // [2019-03-10T03:00:00-06:00, 2019-03-10T04:00:00-06:00)
+    // [2019-03-10T04:00:00-06:00, 2019-03-10T06:00:00-06:00)
+    // ... skipped ...
+    // [2019-11-03T00:00:00-06:00, 2019-11-03T02:00:00-07:00)
+    // [2019-11-03T02:00:00-07:00, 2019-11-03T04:00:00-07:00)
+    // [2019-11-03T03:00:00-07:00, 2019-11-03T06:00:00-07:00)
+
+    intervals(every:1h, location: "America/Denver")
+    // [2019-03-10T00:00:00-07:00, 2019-03-10T01:00:00-07:00)
+    // [2019-03-10T01:00:00-07:00, 2019-03-10T03:00:00-06:00)
+    // [2019-03-10T03:00:00-06:00, 2019-03-10T04:00:00-06:00)
+    // [2019-03-10T04:00:00-06:00, 2019-03-10T05:00:00-06:00)
+    // ... skipped ...
+    // [2019-11-03T00:00:00-06:00, 2019-11-03T01:00:00-06:00)
+    // [2019-11-03T01:00:00-06:00, 2019-11-03T02:00:00-07:00)
+    // [2019-11-03T02:00:00-07:00, 2019-11-03T03:00:00-07:00)
+    // [2019-11-03T03:00:00-07:00, 2019-11-03T04:00:00-07:00)
+
+    intervals(every:30m, location: "America/Denver")
+    // [2019-03-10T00:00:00-07:00, 2019-03-10T00:30:00-07:00)
+    // [2019-03-10T00:30:00-07:00, 2019-03-10T01:00:00-07:00)
+    // [2019-03-10T01:00:00-07:00, 2019-03-10T01:30:00-07:00)
+    // [2019-03-10T01:30:00-07:00, 2019-03-10T03:00:00-06:00)
+    // [2019-03-10T03:00:00-06:00, 2019-03-10T03:30:00-06:00)
+    // [2019-03-10T03:30:00-06:00, 2019-03-10T04:00:00-06:00)
+    // ... skipped ...
+    // [2019-11-03T00:00:00-06:00, 2019-11-03T00:30:00-06:00)
+    // [2019-11-03T00:30:00-06:00, 2019-11-03T01:00:00-06:00)
+    // [2019-11-03T01:00:00-06:00, 2019-11-03T01:30:00-06:00), [2019-11-03T01:00:00-07:00, 2019-11-03T01:30:00-07:00)
```

Comment:
This was the weirdest thing. When I started trying to get some edge cases that tripped up influxql, I came out with this as the most logical conclusion. Since we automatically adjust a window to add or remove an hour for week long intervals, it made sense to do it in partial amounts for shorter intervals. And when the interval was shorter than the time offset, it made sense to duplicate it multiple times.

I'm not sure that this is sustainable though so we may have to discuss how to deal with this. I'm not sure how you would deal with a minute interval when it comes to daylight savings time although I guess it makes sense that each minute happens twice during that hour. I am concerned about what the proper `_start` and `_stop` values would be though for that time period.

Reply:
On a positive note, they do appear to be distinct time ranges so as long as it is ok for window to create overlapping ranges, which to my knowledge it is, we can use the min start range as the start and the max end range as the end. We will just have to update the bounds object to understand that start and stop aren't the only things that are important.

Reply:
UTC is monotonic, meaning it always increases. Given a time in nanoseconds since the epoch plus an offset you can always map that back to a UTC time. 

Conversely you cannot always map a UTC time to a timestamp with an offset since there are multiple time/offset combinations that represent the same instant in time. 

Using UTC to handle windowing removes the ambiguity of _duplicate_ windows because each window has distinct start and stop times.

So if we can map a time with its location to a UTC time we can both determine the _start and _stop boundaries of a window as well as determine which times belong in which windows.

Mapping a single time with its location to a UTC time is straightforward.

The only remaining question is how to add a duration to a time with a location when the offset changes over that duration.

I don't follow how that is defined with these examples:

For the `1d` example it seems we change the duration to be longer or shorter such that the time of day remains constant. `2019-03-10T00:00:00-07:00 + 1d =  2019-03-11T00:00:00-06:00` in `America/Denver`

For the `1h` example is seems that we ignore the change in offset. `2019-03-10T01:00:00-07:00 + 1h = 2019-03-10T03:00:00-06:00`  in `America/Denver`.

Can you clarify the behavior of adding a duration to a time with a location where the duration spans a period where the offset changes?

Reply:
My goal was that if the location is set, adding a duration isn't the same as "adding a group of nanoseconds". Which meant that if I added an hour to a time, my next time should be `hour + 1`. I based this off of the idea that if I group by a week, the week should always start on the same day. In order to make that true, then we need a definition where the extra hour from going off of DST is part of that week rather than ignored and going on DST removes that hour.

> For the 1h example is seems that we ignore the change in offset. `2019-03-10T01:00:00-07:00 + 1h = 2019-03-10T03:00:00-06:00` in `America/Denver`.

I am not sure what you mean here. If you add one hour to 1 AM before DST starts, you get 2 AM. But 2 AM when DST starts is actually 3 AM because, at exactly 2 AM, the hour springs forward. That's why it lists 3 AM as the end time for this interval. Another way to write this would be that the range is `[2019-03-10T01:00:00-07:00, 2019-03-10T01:59:59.999999999-07:00]`.

Reply:
I don't know if I am oversimplifying things or if I am missing the point. If so, completely ignore this comment 😅 .

@nathanielc 

> The only remaining question is how to add a duration to a time with a location when the offset changes over that duration.

I would say that a time is always related to a location (if not specified is the local one), so, if you add a duration to a time, you get a time in that precise location (again).

Say you have:

`2019-03-10T00:00:00 @ America/Denver`

This resolves to 

`2019-03-10T00:00:00-XXX`

As far as I  got it, `-XXX` depends on both on the time and the location and can be extracted from a global db.  
For example, `-XXX` is `-07:00` because it's the tenth of March and we are in Denver.

Anyways, now you resolved `@America/Denver` to `-07:00`.

And you can get the UTC time out of it:

`2019-03-10T00:00:00-07:00 = 2019-03-09T17:00:00Z`

So, now you can add 1 day:

`2019-03-09T17:00:00Z + 1d = 2019-03-10T17:00:00Z`

And you can come back to Denver that would now resolve to `-06:00` thanks to the database:

`2019-03-10T23:00:00-06:00`

Is the computation correct in your opinion? 🤔 

By the end of the day we have: 

`2019-03-10T00:00:00 @ America/Denver + 1d = 2019-03-10T23:00:00 @ America/Denver`

If you suppose that the timezone changes exactly at midnight, then you'll obtain that:

`2019-03-10T00:00:00 @ America/Denver + 1d == 2019-03-10T00:00:00 @ America/Denver + 1d1h`

Because, in Denver, it's twice `2019-03-10T23:00:00`. But that seems coherent with what happens in reality 🤔 

Another thing I wanted to say, is that I think that here we can employ a new operator and use `@` as suggested by @nathanielc in a previous PR. That seems idiomatic and has the pro that it can be applied to any time without adding parameters to every function that works with time.

Reply:
> And you can come back to Denver that would now resolve to -06:00 thanks to the database:
>
> 2019-03-10T23:00:00-06:00
>
> Is the computation correct in your opinion? 🤔

I believe this behavior is incorrect. After you resolve back to Denver, you note that the offset changed and so you need to now adjust the unix timestamp as if the start time had been in the -6 offset instead of -7. This is to keep it so that adding a day always results in midnight being the time.

About the operator, I was thinking that we would remove the ability to use an operator to add times. I was talking about this with @jlapacik at some point and it might not be possible for us to allow different types on the left and right side of the parameter. If this is the case and adding a time isn't too common to begin with, I think we can add functions for performing these operations to the `date` package. I think it would probably be easier than adding another operator token.

Reply:
>This is to keep it so that adding a day always results in midnight being the time.

I think this makes sense for durations of a day, but durations less than a day it starts to become ambiguous.  

For example if we apply this same reasoning to hours you state this:

> Which meant that if I added an hour to a time, my next time should be hour + 1.

Makes sense, if I add an hour to a date time I should get a date time that has its `hour` unit incremented by one. 

But then its not really possible to do that as some hours of the day simply do not exist

> If you add one hour to 1 AM before DST starts, you get 2 AM. But 2 AM when DST starts is actually 3 AM because, at exactly 2 AM, the hour springs forward. That's why it lists 3 AM as the end time for this interval.

So if I add 1 hour to `1 AM` I would expect to get `2 AM` but I can't because `2 AM` doesn't exist in that location. 

As a result is seems like it is impossible to preserve the invariant that `hour + 1` is your hour later according to the adjusted timezone.

I think that preserving midnight for day durations is a reasonable invariant to keep because you can always add a `day` unit to a date and have a next day that makes sense. The duration may not always be exactly 24hours but there is always a date to represent the new value.


In light of both of these I think we want to do two things to make this much easier to reason about:

1. Reintroduce `day` as its own base unit. Define day units as not strictly 24h but always the next day at the same time of day after adjusting for changes in offsets.
2. Strictly define the nanoseconds unit as adding that number of nanoseconds always regardless of offset changes.

This way we have three base units defined as follows:

1. Month, adding months always increases the month number accordingly. If a date would go past the end of the month then the last day of the month is used.
2. Day, adding days always increases the days number accordingly and the time of day will always be the same as the starting date including when the offset changes.
3. Nanosecond, adding nanoseconds always move the date time forward in time by that exact number of nanoseconds regardless of timezone changes.


The point here is we make it easy for users to reason about how time/duration addition works. 

Thoughts?

Reply:
I guess I am concerned about offset changes that are not related to daylight savings time which is why I had favored keeping the definition away from defining a particular offset shift. And I wonder what would happen if you wanted to create a window interval for every 8 hours to signify morning, afternoon, and night. This may be common if you were trying to organize data around [work shift hours](https://www.ziprecruiter.com/blog/what-are-shift-work-hours/). It might be possible to do this by creating 3 intervals and using day increases to keep it aligned, but I'm not sure that's very user friendly and I'd like us to figure out the performance implications of binding ourselves to that method.

So I don't really think defining day as a part of the vector really helps except for one specific category of problem with one specific interval that may end up being US-centric because not every location deals with time zones in the hour forward/hour back at the specific time of 2 AM.

It might be useful if we created a research issue to look at time zones and see if there's any prior art for this.

> So if I add 1 hour to 1 AM I would expect to get 2 AM but I can't because 2 AM doesn't exist in that location.
>
> As a result is seems like it is impossible to preserve the invariant that hour + 1 is your hour later according to the adjusted timezone.

I mentioned it before but the original idea that InfluxQL was supposed to implement, but I messed up with some edge cases, was that this invariant would only get preserved if the shift was less than the interval. So if you had an hour shift from daylight savings time, then grouping by an hour or less wouldn't cause it to worry about daylight savings time. If you had an interval larger than that, we would automatically add/remove time values from the interval to accommodate. This would solve the problem of adding a day and it would also allow Flux to be used for creating intervals for people's working hours. It may match more with how daylight savings time is practiced in real life where if you are working a night shift when the clock shifts, you work either an hour more or less during that night.

Reply:
> It might be useful if we created a research issue to look at time zones and see if there's any prior art for this.

I have been looking at existing systems this is what I have found:

* Go - treats durations a constant lengths of time regardless of timezone
* Python - same as Go
* Excel/Lbre Calc - Most operations work with only dates or times but not both. Operations that work for both dates and times are not timezone aware (or at least I didn't find any).
* R - Has lots of libraries that do things differently. The only library I found that has support for timezones is this one https://lubridate.tidyverse.org/ Its worth a more complete discussion.

The Lubridate library in R is inspired by this Java lib https://www.joda.org/. Lubridate has three different types for working with durations of time, from their site:

* durations, which measure the exact amount of time between two points
 * periods, which accurately track clock times despite leap years, leap seconds, and day light savings time
 * intervals, a protean summary of the time information between two points

This cheat sheet (second page) is particularly insightful https://rawgit.com/rstudio/cheatsheets/master/lubridate.pdf

In short I have been arguing for making Flux durations behave like the Lubridate durations. It seems like @jsternberg has been arguing for Flux durations to behave like Lubridate periods.

Perhaps we should just create two different duration types? 

FWIW I still think my proposal above will work for most cases as it is still possible to describe periods of time even in the face of time zone shifts since you can always anchor a start time at midnight for the time zone and use offsets from there.

Reply:
Yes, I think periods are more like what we want. Although I would say your proposal doesn't behave like durations because I think your proposal is that it will behave like durations if you don't use days and it will behave like a period if you use a day. That's the part that seems to be weird for me.

I think the Lubridate library is a perfect model. Although I guess the reason why I think we should operate like periods in all circumstances is that, if you want durations, you can just use UTC and avoid time zones entirely. If you want a specific offset, use an offset. If you're setting a time zone, I think that the user has indicated they want to track clock times. I think this part of the document says it best:

> R recognizes ~600 time zones. Each encodes the time zone, Daylight
Savings Time, and historical calendar variations for an area. R assigns
one time zone per vector.
>
> Use the UTC time zone to avoid Daylight Savings.

I am curious what this library does if you add 1 minute repeatedly at a DST switch while in a time zone. It seems like that would be the solution to my weird intervals.